### ISSUE TITLE
`pmove_float` causes acceleration on slick with no user input.

### ISSUE THREAD

**Issue by oitzujoey** [OpenArena/gamecode#141]
Go to a map that has slick, such as psidm7 or dfwc2017-6.

Set these settings:
```
pmove_float 0
pmove_fixed 1
```
Step on the slick. Nothing of interest happens.

```
pmove_float 1
pmove_fixed 0
```
Step on the slick. You begin to slowly accelerate.

I see one call to `trap_SnapVector` in the code, so my guess is that this bug is present in Q3 as well, but hidden by snapping.

**Comment by oitzujoey** [comment]
It turns out that this is related to the overbounce bug.
`code/game/bg_pmove.c:815`
```c
if ( ( pml.groundTrace.surfaceFlags & SURF_SLICK ) || pm->ps->pm_flags & PMF_TIME_KNOCKBACK ) {
	// Gravity is added to allow slope physics to work properly on slick surfaces.
	pm->ps->velocity[2] -= pm->ps->gravity * pml.frametime;
}

// Save speed, which has just had some downward velocity added to it.
vel = VectorLength(pm->ps->velocity);

// Clip that downward velocity that we added.
PM_ClipVelocity (pm->ps->velocity, pml.groundTrace.plane.normal, 
	pm->ps->velocity, OVERCLIP );

// don't decrease velocity when going up or down a slope
// I'm not sure why this is here, but we should probably keep it. Was this added for slick?
VectorNormalize(pm->ps->velocity);
VectorScale(pm->ps->velocity, vel, pm->ps->velocity);
```
`-pm->ps->gravity * pml.frametime` accelerates the player downwards, but most or all of it is clipped by `PM_ClipVelocity`. The velocity's speed is set to the original speed, so much of the velocity from gravity is redirected horizontally and the player accelerates.

One way of fixing this is to move the gravity code below the overbounce code and then add another `PM_ClipVelocity` after that.
```c
vel = VectorLength(pm->ps->velocity);

// slide along the ground plane
PM_ClipVelocity (pm->ps->velocity, pml.groundTrace.plane.normal, 
	pm->ps->velocity, OVERCLIP );

// don't decrease velocity when going up or down a slope
VectorNormalize(pm->ps->velocity);
VectorScale(pm->ps->velocity, vel, pm->ps->velocity);

if ( ( pml.groundTrace.surfaceFlags & SURF_SLICK ) || pm->ps->pm_flags & PMF_TIME_KNOCKBACK ) {
	pm->ps->velocity[2] -= pm->ps->gravity * pml.frametime;

	// Extra clip for slick gravity.
	PM_ClipVelocity (pm->ps->velocity, pml.groundTrace.plane.normal, 
		pm->ps->velocity, OVERCLIP );
} else {
	// don't reset the z velocity for slopes
//	pm->ps->velocity[2] = 0;
}
```
This does change the way movement works on slopes, but hopefully not too noticeably. If necessary, this could be changed only when `pmove_float` is set.

**Comment by The-Gig** [comment]
If I understood corectly, the name of this issue should be changed from "_pmove_fixed_ causes acceleration..." to "_pmove_float_ causes acceleration...", right?

**Comment by sago007** [comment]
Can you explain where in psidm7 can this be observed?
I ran around the map trying to find a place to test the difference but I could not spot a difference.

**Comment by oitzujoey** [comment]
> If I understood corectly, the name of this issue should be changed from "_pmove_fixed_ causes acceleration..." to "_pmove_float_ causes acceleration...", right?

You are correct.

**Comment by oitzujoey** [comment]
> Can you explain where in psidm7 can this be observed?
> I ran around the map trying to find a place to test the difference but I could not spot a difference.

There are two places. One is in the RA room. Opposite the RA is an elevated tunnel. RJ on top of the tunnel to find that slick surface.
If you go through the tunnel and come out, you will find another slick if you RJ on top of the exit.

**Comment by The-Gig** [comment]
Slick surfaces are quite rare I guess... In baseoa 0.8.8 it looks like the only shader with "surfaceparm slick" is common/slick (from common.shader), and I found no stock .bsp file mentioning "slick".

Anyway, I can guess there might be some third party snowy/icy map out there _featuring it to some noticeable degree_, while I guess having "slick" on large parts of the map would make it very hard to play.

I vaguely recall trying to use it years ago, I tried adding it to the aqueduct of oacmpdm10, but at the end I left it out, probably also due to this bug (I'm not sure, maybe it can even make you slide _upwards_ if placed on a slope not steep enough! My idea was to make the aqueduct slightly sloped and slick to give the idea of the water flow pushing you downwards, but it didn't go well and I discarded the idea).

**Comment by NeonKnightOA** [comment]
am_icyhell2 and am_icyhell3 IICR make use of that kind of surfaces. I don't remember if it's really the case, though.

**Comment by The-Gig** [comment]
Sago, I noticed this commit: https://github.com/OpenArena/gamecode/commit/880d450a16695a1bdcf32ccb125b3a20d3930383
I guess you are testing the "ignore float when on a slick surface" solution. Does that mean that when a player/bot will be on a slick surface with pmove_float enabled, his movement will temporarily work as "fps-dependent" or "fixed" depending from pmove_fixed being 0 or 1?

**Comment by sago007** [comment]
Yes, that is the point.
It is the simplest solution.
I do not think it is going to be that big of a deal. FPS dependent physics are mostly relevant while airborne because of gravity in other cases the framerate physics will average out.

**Comment by sago007** [comment]
Merged to master.
This should handle this case.

**Comment by The-Gig** [comment]
Tested, modifying a simple test map for the need.
What happens without touching the controls anymore.

Pmove_float 0:
slick levelled surface: you keep the same speed
slick sloped surface (going up): you keep the same speed
slick sloped surface (going down): you keep increasing speed

Pmove_float 1:
slick levelled surface: OA 0.8.8 you keep increasing speed; OAX 2021_09_26 you keep the same speed
slick sloped surface (going up): OA 0.8.8 you slowly slow down, up to the point you get some lateral movement and then start going backwards (down); OAX 2021_09_26 you keep the same speed
slick sloped surface (going up): you keep increasing speed

So, the problem I did remember with "sliding upwards" happened with move_float 0, not with pmove_float 1. _In that specific case_ (going upwards on a sloped surface) the 0.8.8 behavior "made more sense", as it did something more similar to the real world. But I can guess that it might have been considered "altering the Q3 (odd) behavior"?
